src/emulate.c

```diff
@@ -50,41 +50,43 @@ enum {
 #undef _
 };
 
-static void rv_exception_default_handler(struct riscv_t *rv)
+static void rv_exception_default_handler(struct riscv_t *rv,
+                                         struct rv_insn_t *ir)
 {
-    rv->csr_mepc += rv->insn_len;
+    rv->csr_mepc += ir->insn_len;
     rv->PC = rv->csr_mepc; /* mret */
 }
 
-#define EXCEPTION_HANDLER_IMPL(type, code)                                \
-    static void rv_except_##type(struct riscv_t *rv, uint32_t mtval)      \
-    {                                                                     \
-        /* mtvec (Machine Trap-Vector Base Address Register)              \
-         * mtvec[MXLEN-1:2]: vector base address                          \
-         * mtvec[1:0] : vector mode                                       \
-         */                                                               \
-        const uint32_t base = rv->csr_mtvec & ~0x3;                       \
-        const uint32_t mode = rv->csr_mtvec & 0x3;                        \
-        /* mepc  (Machine Exception Program Counter)                      \
-         * mtval (Machine Trap Value Register)                            \
-         * mcause (Machine Cause Register): store exception code          \
-         */                                                               \
-        rv->csr_mepc = rv->PC;                                            \
-        rv->csr_mtval = mtval;                                            \
-        rv->csr_mcause = code;                                            \
-        if (!rv->csr_mtvec) { /* in case CSR is not configured */         \
-            rv_exception_default_handler(rv);                             \
-            return;                                                       \
-        }                                                                 \
-        switch (mode) {                                                   \
-        case 0: /* DIRECT: All exceptions set PC to base */               \
-            rv->PC = base;                                                \
-            break;                                                        \
-        /* VECTORED: Asynchronous interrupts set PC to base + 4 * code */ \
-        case 1:                                                           \
-            rv->PC = base + 4 * code;                                     \
-            break;                                                        \
-        }                                                                 \
+#define EXCEPTION_HANDLER_IMPL(type, code)                                 \
+    static void rv_except_##type(struct riscv_t *rv, struct rv_insn_t *ir, \
```

Comment:
Can you avoid adding `ir` a a parameter since we can look up the translated instruction from `rv`?

Reply:
Yes, use queue to mange memory block. At the same time, the block occurred the exception always at the "rear".